# ML-03 — Frame Your Lane as an ML Task

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Rislantrs/flyrank-ml-internship/blob/main/work/notebooks/w02_ml_task_framing.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*

In [5]:
import pandas as pd

# URL mentah ke dataset di GitHub
github_csv_url = 'https://raw.githubusercontent.com/Rislantrs/flyrank-ml-internship/main/data/raw/content_refresh_anonymized.csv'

# Memuat data untuk melihat seberapa banyak konten yang akan di-ranking
df = pd.read_csv(github_csv_url)

# Menampilkan jumlah total halaman web yang akan dievaluasi dan diberikan skor prioritas
total_content = df['content_id'].nunique()
print(f"Total konten unik yang perlu diberikan skor prioritas (di-ranking): {total_content} halaman")

# Menampilkan beberapa baris pertama data untuk pemeriksaan cepat
print("\nBeberapa baris pertama data:")
display(df.head())

Total konten unik yang perlu diberikan skor prioritas (di-ranking): 30000 halaman

Beberapa baris pertama data:


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,provider_used,model_used,impressions_90d,clicks_90d,pageviews_90d,sessions_90d,users_90d,engaged_sessions_90d,ai_sessions_90d,scroll_events_90d,days_with_impressions,days_with_sessions,impressions_last_30d,clicks_last_30d,sessions_last_30d,impressions_prev_30d,clicks_prev_30d,sessions_prev_30d,content_age_days,age_tier,age_tier_order,days_since_last_update,freshness_tier,word_count_tier,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,NaN,gemini-2.5-flash,3803,29,22,17,16,1,0,1,88,13,578,2,2,987,13,9,187,181-365,5,20,0-30,2000-3500,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,NaN,gemini-3-flash-preview,15320,7,10,9,9,0,0,1,88,9,2501,2,3,5915,1,2,445,365+,6,25,0-30,2000-3500,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,NaN,gemini-2.5-flash,12581,11,14,11,11,0,0,4,88,11,2382,1,1,6089,3,3,141,91-180,4,20,0-30,3500+,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,NaN,NaN,11751,58,87,78,75,1,0,3,88,51,3626,22,35,4206,17,26,463,365+,6,22,0-30,NaN,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,NaN,gemini-3-flash-preview,19140,24,177,145,144,0,0,43,88,33,4211,10,14,6452,2,9,263,181-365,5,14,0-30,2000-3500,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [4]:
print("Daftar semua fitur (kolom) dalam dataset:")
display(df.columns.tolist())

Daftar semua fitur (kolom) dalam dataset:


['content_id',
 'client_id',
 'search_volume',
 'competition',
 'competition_level',
 'cpc',
 'content_type',
 'main_intent',
 'word_count',
 'char_count',
 'provider_used',
 'model_used',
 'impressions_90d',
 'clicks_90d',
 'pageviews_90d',
 'sessions_90d',
 'users_90d',
 'engaged_sessions_90d',
 'ai_sessions_90d',
 'scroll_events_90d',
 'days_with_impressions',
 'days_with_sessions',
 'impressions_last_30d',
 'clicks_last_30d',
 'sessions_last_30d',
 'impressions_prev_30d',
 'clicks_prev_30d',
 'sessions_prev_30d',
 'content_age_days',
 'age_tier',
 'age_tier_order',
 'days_since_last_update',
 'freshness_tier',
 'word_count_tier',
 'char_count_tier',
 'ctr',
 'avg_position',
 'engagement_rate',
 'scroll_rate',
 'ai_traffic_pct',
 'impression_tier',
 'position_tier',
 'trend_direction',
 'trend_pct']

The primary focus of this problem solving is to optimize content team decisions and workflows. Instead of simply classifying whether content will perform poorly (yes/no), the ranking approach will evaluate historical data over the past 90 days to generate a priority score for each webpage.

The output, in the form of a sorted content list (ranked queue), will provide actionable insights for editors: which content has the greatest impact and is most urgent to update or fix first.

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Memeriksa nilai apa saja yang ada di kolom trend_direction
print("Distribusi Arah Tren Konten:")
print(df['trend_direction'].value_counts(dropna=False))

# Memperbaiki nama target sesuai dengan data yang sebenarnya
target_kondisi = (df['trend_direction'] == 'down')
print(f"Jumlah konten yang menurun: {target_kondisi.sum()} halaman")


Distribusi Arah Tren Konten:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64
Jumlah konten yang menurun: 16262 halaman


The target to be predicted is the probability of performance decline, which is derived from the `trend_direction` column. We convert it into a binary target: 1 if 'down' (declining) and 0 for other categories (stable, up, flat, new).

The model will output a probability score (e.g., 90% chance of being 'down'). This probability score will be our proxy for ranking the content queue, from the most urgent to be fixed.

Note: Since `trend_direction` and `trend_pct` are direct observations (ground truth), these two columns MUST be removed from the features (input) to prevent data leakage.

## 3. Success metric

*One metric you can defend. What number means 'good'?*

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Menghitung Base Rate (Tebakan Acak) sebagai Baseline untuk Precision@K
total_halaman = len(df)
jumlah_down = target_kondisi.sum()

base_rate = (jumlah_down / total_halaman) * 100

print(f"Base Rate (Peluang acak halaman berstatus 'down'): {base_rate:.2f}%")
print("Artinya, jika tanpa ML kita mengambil 100 halaman secara acak, presisinya hanya ~54%.")
print("Target Metrik: Precision@100 dari model ML kita harus jauh lebih tinggi dari 54% (idealnya di atas 80-90%).")

Base Rate (Peluang acak halaman berstatus 'down'): 54.21%
Artinya, jika tanpa ML kita mengambil 100 halaman secara acak, presisinya hanya ~54%.
Target Metrik: Precision@100 dari model ML kita harus jauh lebih tinggi dari 54% (idealnya di atas 80-90%).


The main success metric for this ranking system is Precision@K (e.g., Precision@100 or Precision@500).

Business reason: The editorial team has limited capacity. If we provide a queue of the top 100 contents (K=100) predicted to be the most urgent to fix, we want to ensure that most (close to 100%) of that list are indeed contents with declining performance.

As a secondary metric, we will use ROC-AUC to ensure the model's overall ability to separate (rank probabilities) content that is declining versus not declining. The model is considered 'good' if its Precision@K and ROC-AUC significantly outperform the base rate (random guess), which we calculated to be **54.21%** for pages with 'down' status.

## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Memilih beberapa kolom inti untuk menunjukkan wujud "Unit Analisis" kita
kolom_inti = ['content_id', 'client_id', 'content_type', 'clicks_90d', 'trend_direction']
df_unit = df[kolom_inti].head()

print("Bisa dilihat di bawah, 1 baris (index) = 1 ID Konten:")
display(df_unit)

# Memastikan tidak ada duplikasi (1 baris benar-benar 1 konten unik)
print(f"\nJumlah baris data: {len(df)}")
print(f"Jumlah content_id unik: {df['content_id'].nunique()}")
print("Karena jumlah baris dan ID unik sama, terbukti unit analisis kita adalah per-konten.")

Bisa dilihat di bawah, 1 baris (index) = 1 ID Konten:


,content_id,client_id,content_type,clicks_90d,trend_direction
0,content_304f48230142,client_f369cb89fc,keyword article,29,down
1,content_a1fb4e703a9e,client_4e07408562,keyword article,7,down
2,content_9aa793d4d895,client_7f2253d7e2,keyword article,11,down
3,content_331d6c4de07b,client_19581e27de,keyword article,58,stable
4,content_d99b7a2d90ca,client_3fdba35f04,keyword article,24,down



Jumlah baris data: 30000
Jumlah content_id unik: 30000
Karena jumlah baris dan ID unik sama, terbukti unit analisis kita adalah per-konten.


Unit analisis dalam dataset ini adalah satu item konten (halaman web) milik satu klien tertentu.

Satu baris data mewakili akumulasi performa (metrik SEO dan engagement) dari satu halaman web tersebut yang diukur selama periode 90 hari terakhir.

The unit of analysis in this dataset is a single content item (webpage) belonging to a specific client.

One row of data represents the accumulated performance (SEO and engagement metrics) of that one webpage, measured over the last 90 days.

## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*

In [12]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Menunjukkan kompleksitas data:
# Mengabaikan aturan sederhana dengan melihat interaksi beberapa fitur terhadap status 'down'

features_to_check = ['clicks_90d', 'avg_position', 'scroll_rate', 'word_count']

print("Rata-rata metrik berdasarkan status down:")
summary = df.groupby(target_kondisi)[features_to_check].mean()
display(summary.round(2))

print("\nPenjelasan Kode:")
print("Tabel di atas menunjukkan bahwa konten yang 'down' memiliki nilai rata-rata yang saling bertumpuk dengan yang 'tidak down'.")
print("Satu aturan sederhana (misal: jika clicks_90d < X) pasti akan salah mengklasifikasikan banyak halaman. ML dibutuhkan untuk memproses gabungan seluruh sinyal ini.")

Rata-rata metrik berdasarkan status down:


,clicks_90d,avg_position,scroll_rate,word_count
trend_direction,,,,
False,19.34,16.82,18.31,2957.45
True,13.36,15.94,18.13,3221.83



Penjelasan Kode:
Tabel di atas menunjukkan bahwa konten yang 'down' memiliki nilai rata-rata yang saling bertumpuk dengan yang 'tidak down'.
Satu aturan sederhana (misal: jika clicks_90d < X) pasti akan salah mengklasifikasikan banyak halaman. ML dibutuhkan untuk memproses gabungan seluruh sinyal ini.


Using rigid rules (like if-else statements) is not sufficient to prioritize content updates because the patterns of performance decline are highly complex and involve the interaction of dozens of signals simultaneously (clicks, Google position, scroll rate, content type, and even AI search performance).

Rigid rules would impose thresholds that are too generalized for all clients, whereas the 'decline' threshold for a small blog differs from that of a corporate website. Machine Learning excels here because it can monitor non-linear interactions of 40+ features simultaneously, providing subtle (continuous) probability estimates to rank priorities much more accurately than manual if-else rules.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.